File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
dic = {
    ")":"(",
    "]":"["
}
def input_check():
    data = input()

    if data.endswith("."):
        return data

while True:
    data = input_check()
    if data == '.':
        break
    stack = []
    flag = True
    for val in data:
        if val == "(" or val == "[":
            stack.append(val)
        elif (val == ")" or val == "]") and stack:
            if stack[-1] == dic[val]:
                stack.pop()
            else:
                break
        elif val == ")" or val == "]"and len(stack) == 0:
            flag = False
            break

    if flag == False or stack:
        print("no")
    else:
        print("yes")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Map<String, String> dic = new HashMap<>();
        dic.put(")", "(");
        dic.put("]", "[");
        
        Scanner inner = new Scanner(System.in);
        while(true){
            String sentence = inner.nextLine();
            if (sentence.equals(".")){
                break;
            }

            Stack stack = new Stack();
            Boolean flag = true;
            for (String alpha: sentence.split("")) {
                if (alpha.equals("(") || alpha.equals("[")){
                    stack.push(alpha);
                } else if (alpha.equals(")") || alpha.equals("]")) {
                    if (stack.isEmpty()){
                        flag = false;
                        break;
                    }
                    if (stack != null && dic.get(alpha).equals(stack.peek())){
                        stack.pop();
                    }else{
                        flag = false;
                        break;
                    }
                }
            }

            if (flag == false || !stack.isEmpty()){
                System.out.println("no");
            }else{
                System.out.println("yes");
            }
        }
    }
}